utilities/blob_db/blob_db_impl.cc

```diff
@@ -926,15 +926,41 @@ uint64_t BlobDBImpl::ExtractExpiration(const Slice& key, const Slice& value,
   return has_expiration ? expiration : kNoExpiration;
 }
 
+std::shared_ptr<BlobFile> BlobDBImpl::GetOldestBlobFile() {
+  std::vector<std::shared_ptr<BlobFile>> blob_files;
+  CopyBlobFiles(&blob_files);
+  if (blob_files.empty()) {
+    return nullptr;
+  }
+  blobf_compare_ttl compare;
+  return *std::min_element(blob_files.begin(), blob_files.end(), compare);
+}
+
+bool BlobDBImpl::EvictOldestBlobFile() {
+  auto oldest_file = GetOldestBlobFile();
+  if (oldest_file == nullptr) {
+    return false;
+  }
+
+  WriteLock wl(&mutex_);
+  oldest_file->SetCanBeDeleted();
+  obsolete_files_.push_front(oldest_file);
+  return true;
+}
+
 Status BlobDBImpl::AppendBlob(const std::shared_ptr<BlobFile>& bfile,
                               const std::string& headerbuf, const Slice& key,
                               const Slice& value, uint64_t expiration,
                               std::string* index_entry) {
   auto size_put = BlobLogRecord::kHeaderSize + key.size() + value.size();
-  if (bdb_options_.blob_dir_size > 0 &&
-      (total_blob_space_.load() + size_put) > bdb_options_.blob_dir_size) {
-    if (!bdb_options_.is_fifo) {
-      return Status::NoSpace("Blob DB reached the maximum configured size.");
+  uint64_t current_space_util = total_blob_space_.load();
+  if (bdb_options_.blob_dir_size > 0) {
+    if (current_space_util + size_put > bdb_options_.blob_dir_size) {
```

Comment:
* Should also trigger `EvictOldestBlobFile` when out of space
* total_blob_space_ includes obsolete files size. We need another counter which excludes obsolete files.
* total_blob_space_ doesn't include file head/footer size. Should add the count on file close.

Reply:
`EvictOldestBlobFile` already gets called at 90% full and if `is_fifo` is set. So its not necessary to call here. Also note that the execution shouldn't generally get here if `is_fifo = true`.

Reply:
There's a rare case when the db jump from 80% full to 100% full. Also I would prefer we don't return NoSpace when is_fifo = true.

Reply:
True, I agree. will make the change.

Reply:
I did not make changes to the total_blob_space_ computation. I'll make those changes in a later PR.